lib/bson/decimal128.rb

```diff
@@ -0,0 +1,298 @@
+
+    # The native type to which this object can be converted.
+    #
+    # @since 4.1.0
+    NATIVE_TYPE = BigDecimal.freeze
+
+    # Get the Decimal128 as JSON hash data.
+    #
+    # @example Get the Decimal128 as a JSON hash.
+    #   decimal.as_json
+    #
+    # @return [ Hash ] The number as a JSON hash.
+    #
+    # @since 4.1.0
+    def as_json(*args)
+      { EXTENDED_JSON_KEY => to_s }
+    end
+
+    # Check equality of the decimal128 object with another object.
+    #
+    # @example Check if the decimal128 object is equal to the other.
+    #   decimal == other
+    #
+    # @param [ Object ] other The object to check against.
+    #
+    # @return [ true, false ] If the objects are equal.
+    #
+    # @since 4.1.0
+    def ==(other)
+      return false unless other.is_a?(Decimal128)
+      @high == other.instance_variable_get(:@high) &&
+          @low == other.instance_variable_get(:@low)
+    end
+    alias :eql? :==
+
+    # Create a new Decimal128 from a BigDecimal.
+    #
+    # @example Create a Decimal128 from a BigDecimal.
+    #   Decimal128.new(big_decimal)
+    #
+    # @param [ BigDecimal ] big_decimal The BigDecimal to use for
+    #   instantiating a Decimal128.
+    #
+    # @raise [ InvalidBigDecimal ] Raise error unless object argument is a BigDecimal.
+    #
+    # @since 4.1.0
+    def initialize(big_decimal)
+      raise InvalidBigDecimal.new unless big_decimal.is_a?(BigDecimal)
+      set_bits(*Builder::FromBigDecimal.new(big_decimal).bits)
+    end
+
+    # Get the decimal128 as its raw BSON data.
+    #
+    # @example Get the raw bson bytes in a buffer.
+    #   decimal.to_bson
+    #
+    # @return [ BSON::ByteBuffer ] The raw bytes in a buffer.
+    #
+    # @see http://bsonspec.org/#/specification
+    #
+    # @since 4.1.0
+    def to_bson(buffer = ByteBuffer.new, validating_keys = Config.validating_keys?)
+      buffer.put_decimal128(@low, @high)
+    end
+
+    # Get the hash value for the decimal128.
+    #
+    # @example Get the hash value.
+    #   decimal.hash
+    #
+    # @return [ Integer ] The hash value.
+    #
+    # @since 4.1.0
+    def hash
+      num = @high << 64
+      num |= @low
+      num.hash
+    end
+
+    # Get a nice string for use with object inspection.
+    #
+    # @example Inspect the decimal128 object.
+    #   decimal128.inspect
+    #
+    # @return [ String ] The decimal as a string.
+    #
+    # @since 4.1.0
+    def inspect
+      "BSON::Decimal128('#{to_s}')"
+    end
+
+    # Get the string representation of the decimal128.
+    #
+    # @example Get the decimal128 as a string.
+    #   decimal128.to_s
+    #
+    # @return [ String ] The decimal128 as a string.
+    #
+    # @since 4.1.0
+    def to_s
+      @string ||= Builder::ToString.new(self).string
+    end
+    alias :to_str :to_s
+
+    # Get a BigDecimal object corresponding to this Decimal128.
+    # Note that the precision of the resulting BigDecimal is not guaranteed to exactly
```

Comment:
No, because specifying the precision is sometimes lost with BigDecimal objects. For example:

2.2.3 :004 > d = BigDecimal.new(2.00, 3)
 => #<BigDecimal:7fb43c0ed8d0,'0.2E1',9(27)>